README.md
Strowger
========

Simple switching for your Twilio SMS app.

Why
---

[Twilio's SMS capabilities](http://www.twilio.com/sms) are awesome.
[Flask](http://flask.pocoo.org/) apps are awesome.

That 200-line cascade of if/else cases in all of your nontrivial SMS apps?

Not so awesome.

Remembering to return `unicode(response)` from all of your exit points?
Kinda annoying.

Strowger adds dispatch (based on SMS message contents) and TwiML formatting
on top of Flask.

Example
-------

Strowger's basic pattern should look very familiar to you if you've ever
written a Flask application:

    from strowger import Switch

    switch = Switch(__name__)

    @switch.connect('Hello')
    def hello(request, response):
        response.message(msg='Hello, world')
        return response

Strowger matches an incoming SMS body to the
[Python regular expressions](http://docs.python.org/2/library/re.html)
configured in its list of mappings, and invokes the corresponding function.

Handler functions get passed a request object that exposes all of the
fields on [Twilio's TwiML request](http://www.twilio.com/docs/api/twiml/sms/twilio_request),
and an empty `twilio.twiml.Response` object (see https://twilio-python.readthedocs.org/en/latest/usage/twiml.html).

Details
-------

### Routing

Works roughly like Flask routing. Parameters aren't as pretty because we're
matching freeform text rather than URL paths. Named groups are supported
in match patterns, and the match object's `groupdict` will be passed as
keyword objects to the handler function:

    @switch.connect('My name is (?P<name>\w+)')
    def greet(request, response, name):
        response.message(msg='Hello, {}. Nice to meet you.'.format(name))
        return response

`connect()` accepts a `flags` argument to specify behavior for the pattern
matching. (It's passed directly to re.compile.)

    @switch.connect('Goodbye', flags=re.IGNORECASE)
    def farewell(request, response):
        response.message(msg="Good day!")
        return response

By default, Strowger will return a 404 response to any message that doesn't
match a rule. You can change this by passing a handler function to
`app.set_default_handler()`.

### Request Objects

These provide a convenient interface to the elements of the incoming TwiML
request:

- `from_number`: The number the message was sent from.
- `to_number`: The number the message was sent to.
- `message_sid`: The Twilio MessageSid identifying the message.
- `account_sid`: The Twilio AccountSid identifying the Twilio account.
- `message_body`: The full body of the message.
- `media_count`: The number of media items attached to the message.
- `media_items`: A list of `MediaItem` objects with the `content_type`
and `url` of each attached media item.
- `raw_values`: The underlying map of request parameters taken from the Flask
request.
- `flask_request`: If you need to access the underlying Flask request, here
it is. You can always import `flask.request` as well.

### Deployment

Nice job getting this far! Let me know what you're using Strowger for.

The Switch object is a WSGI callable just like a Flask application. (Hint:
it just dispatches straight to the Flask application.)

### Flask Handlers and the Request Path

If your app is simple enough, you probably just have one or two other routes
you want to serve as a plain-old HTTP app. The Flask object is available as
`switch.app` and you can add routes to it as you see fit.

By default, Strowger configures itself as the Flask route for '/' and will
respond to both GET and POST requests. You can override this behavior with
the `request_path` and `methods` keyword arguments to the Switch initializer.

Requirements
------------

* [Flask](http://flask.pocoo.org/)
* [twilio-python](https://github.com/twilio/twilio-python)


CHANGES.md
strowger Changelog
==================

Version 0.1.3
-------------

- Fix bug in TwilioMessageRequest media handling

Version 0.1.2
-------------

Initial release


tests/test_map.py
import re
from mock import Mock
from nose.tools import assert_equal
from unittest import TestCase

from strowger import Map
from strowger import Rule


class MapTestCase(TestCase):
    def setUp(self):
        self.map = Map()
        self.m_handler = Mock()

    def test_init(self):
        assert_equal(self.map.rules, [])
        assert_equal(self.map.default_handler, None)

    def test_add_rule(self):
        self.map.add_rule('foobar', self.m_handler)

        assert_equal(len(self.map.rules), 1)
        rule = self.map.rules[0]
        assert_equal(rule.pattern, re.compile('foobar'))
        assert_equal(rule.handler, self.m_handler)

        m_handler_2 = Mock()
        self.map.add_rule('Quux', m_handler_2, re.IGNORECASE)

        assert_equal(len(self.map.rules), 2)
        rule = self.map.rules[1]
        assert_equal(rule.pattern, re.compile('Quux', re.IGNORECASE))
        assert_equal(rule.handler, m_handler_2)

    def test_match(self):
        self.map.add_rule('foobar', self.m_handler)

        match, handler = self.map.match('foobar')
        assert_equal(match.start(), 0)
        assert_equal(handler, self.m_handler)

    def test_no_match(self):
        self.map.add_rule('quux', self.m_handler)

        match, handler = self.map.match('xyzzy plugh')
        assert_equal(match, None)
        assert_equal(handler, None)

    def test_default_handler(self):
        self.map = Map(default_handler=self.m_handler)
        m_handler_2 = Mock()

        self.map.add_rule('foobar', m_handler_2)

        match, handler = self.map.match('garply')
        assert_equal(match, None)
        assert_equal(handler, self.m_handler)

    def test_match_multiple_rules(self):
        self.map.add_rule('foobar', self.m_handler)
        m_handler_2 = Mock()

        self.map.add_rule('quux', m_handler_2)

        match, handler = self.map.match('foobarbaz')
        assert_equal(match.span(), (0, 6))
        assert_equal(handler, self.m_handler)

        match, handler = self.map.match('quux quuux')
        assert_equal(match.span(), (0, 4))
        assert_equal(handler, m_handler_2)


tests/test_switch.py
import re
from mock import Mock, patch
from nose.tools import assert_equal
from nose.tools import assert_is_instance
from unittest import TestCase

from flask import request
from twilio import twiml

from strowger import switch

TEST_TWILIO_REQUEST_DATA = {
    'From': '+14155551234',
    'To': '+14158675309',
    'NumMedia': '0',
    'MessageSid': 'MM{}'.format('a' * 34),
    'AccountSid': 'AC{}'.format('a' * 34),
}


class SwitchTestCase(TestCase):
    def setUp(self):
        self.flask_patcher = patch('strowger.switch.Flask')
        self.m_flask = self.flask_patcher.start()
        self.map_patcher = patch('strowger.switch.Map')
        self.m_map_class = self.map_patcher.start()

    def tearDown(self):
        self.map_patcher.stop()
        self.flask_patcher.stop()

    def test_init(self):
        m_flask_app = Mock()
        self.m_flask.return_value = m_flask_app

        m_map = Mock()
        self.m_map_class.return_value = m_map

        s = switch.Switch(__name__)

        assert_equal(s.app, m_flask_app)
        m_flask_app.route.assert_called_with('/', methods=['GET', 'POST'])
        assert_equal(s.mapping, m_map)

        m_handler = Mock()
        s = switch.Switch(__name__, default_handler=m_handler)

        self.m_map_class.assert_called_with(default_handler=m_handler)


class SwitchWithFlaskAppTestCase(TestCase):
    def setUp(self):
        self.s = switch.Switch(__name__)
        self.s.app.testing = True
        self.app = self.s.app.test_client()

    def test_basic_handler(self):
        @self.s.connect('foobar')
        def _verify_handler(request, response):
            assert_equal(request.from_number, '+14155551234')
            assert_equal(request.to_number, '+14158675309')
            assert_equal(request.message_body, 'foobarbaz')
            assert_equal(request.media_count, 0)
            assert_is_instance(response, twiml.Response)
            response.message(msg='quux')
            return response


        request_data = TEST_TWILIO_REQUEST_DATA.copy()
        request_data['Body'] = 'foobarbaz'
        response = self.app.post('/', data=request_data)

        assert_equal(response.status_code, 200)

        twiml_response = twiml.Response()
        twiml_response.message(msg='quux')

        assert_equal(response.get_data(), unicode(twiml_response))

    def test_match_with_groups(self):
        @self.s.connect('(?P<first>\w+) (?P<second>\w+)')
        def _verify_handler(request, response, first, second):
            assert_equal(first, 'foo')
            assert_equal(second, 'bar')
            assert_equal(request.message_body, 'foo bar')
            return response

        request_data = TEST_TWILIO_REQUEST_DATA.copy()
        request_data['Body'] = 'foo bar'
        response = self.app.post('/', data=request_data)

        assert_equal(response.status_code, 200)

        twiml_response = twiml.Response()
        assert_equal(response.get_data(), unicode(twiml_response))

    def test_twilio_message_request(self):
        request_data = TEST_TWILIO_REQUEST_DATA.copy()
        request_data.update(
            Body='foo bar',
            NumMedia=2,
            MediaContentType0='image/jpeg',
            MediaUrl0='http://example.com/image0',
            MediaContentType1='image/png',
            MediaUrl1='http://example.com/image1',
            FromCity='San Francisco',
            FromState='CA',
        )

        with self.s.app.test_request_context('/', query_string=request_data):
            twilio_request = switch.TwilioMessageRequest(request)

            assert_equal(twilio_request.message_body, 'foo bar')
            assert_equal(twilio_request.from_number, '+14155551234')
            assert_equal(twilio_request.to_number, '+14158675309')
            assert_equal(twilio_request.message_sid, 'MM{}'.format('a' * 34))
            assert_equal(twilio_request.account_sid, 'AC{}'.format('a' * 34))
            assert_equal(twilio_request.media_count, 2)
            media_items = twilio_request.media_items
            assert_equal(
                media_items[0],
                switch.MediaItem('image/jpeg', 'http://example.com/image0'),
            )
            assert_equal(
                media_items[1],
                switch.MediaItem('image/png', 'http://example.com/image1'),
            )
            assert_equal(twilio_request.flask_request, request)
            assert_equal(twilio_request.raw_values, request.values)


sms.py
import re
from collections import namedtuple

from flask import Flask
from flask import abort
from flask import request
from twilio import twiml


Map = namedtuple('Map', 'pattern handler')
MediaItem = namedtuple('MediaItem', 'content_type url')

RESPONSE_HEADERS = {'Content-Type': 'application/xml'}


class Router(object):

    def __init__(self, default_handler=None):
        self.routes = []
        self.default_handler = default_handler

    def add_route(self, pattern, handler, flags=0):
        compiled = re.compile(pattern, flags=flags)
        self.routes.append(Map(compiled, handler))

    def match(self, body):
        for route in self.routes:
            if route.pattern.match(body):
                return route.handler

        return self.default_handler


class TwilioMessageRequest(object):

    def __init__(self, flask_request):
        self.from_number = flask_request.values['From']
        self.to_number = flask_request.values['To']
        self.message_body = flask_request.values['Body']
        self.media_count = int(flask_request.values['NumMedia'])
        self.media_items = []
        self.raw_values = flask_request.values

        if self.media_count > 0:
            media_items = []
            for index in xrange(self.media_count):
                url = flask_request.values['MediaUrl{}'.format(index)]
                ctype = flask_request.values[
                    'MediaContentType{}'.format(index)
                ]
                media_urls.append(MediaItem(ctype, url))

            self.media_items = media_items


class SmsApp(object):

    def __init__(self, name, request_path='/', default_handler=None, **kwargs):
        '''Create an SmsApp named `name`.

        Additional kwargs are passed to the underlying Flask application.
        '''
        self.app = Flask(name, **kwargs)
        self.router = Router(default_handler=default_handler)

        @self.app.route(request_path, methods=['GET', 'POST'])
        def twilio():
            body = request.values['Body']
            handler = self.router.match(body)

            if handler is None:
                # XXX Is this the right behavior?
                abort(404)

            # Set up our response and request
            twiml_response = twiml.Response()
            twilio_request = TwilioMessageRequest(request)

            try:
                handler(twilio_request, twiml_response)
                return unicode(twiml_response), 200, RESPONSE_HEADERS
            except:
                abort(500)

    def run(self, debug=False):
        self.app.run(debug=debug)


strowger.py
import re
from collections import namedtuple
from functools import wraps

from flask import Flask
from flask import abort
from flask import request
from twilio import twiml


Rule = namedtuple('Rule', 'pattern handler')
MediaItem = namedtuple('MediaItem', 'content_type url')

RESPONSE_HEADERS = {'Content-Type': 'application/xml'}


class Map(object):

    def __init__(self, default_handler=None):
        self.rules = []
        self.default_handler = default_handler

    def add_rule(self, pattern, handler, flags=0):
        compiled = re.compile(pattern, flags=flags)
        self.rules.append(Rule(compiled, handler))

    def match(self, body):
        for rule in self.rules:
            m = rule.pattern.match(body)
            if m is not None:
                return m, rule.handler

        return None, self.default_handler


class TwilioMessageRequest(object):

    def __init__(self, flask_request):
        self.from_number = flask_request.values['From']
        self.to_number = flask_request.values['To']
        self.message_body = flask_request.values['Body']
        self.media_count = int(flask_request.values['NumMedia'])
        self.media_items = []
        self.raw_values = flask_request.values

        if self.media_count > 0:
            media_items = []
            for index in xrange(self.media_count):
                url = flask_request.values['MediaUrl{}'.format(index)]
                ctype = flask_request.values[
                    'MediaContentType{}'.format(index)
                ]
                media_urls.append(MediaItem(ctype, url))

            self.media_items = media_items


class Switch(object):

    def __init__(self, name, request_path='/', default_handler=None, **kwargs):
        '''Create a Switch named `name`.

        Additional kwargs are passed to the underlying Flask application.
        '''
        self.app = Flask(name, **kwargs)
        self.mapping = Map(default_handler=default_handler)

        @self.app.route(request_path, methods=['GET', 'POST'])
        def twilio():
            body = request.values['Body']
            match_obj, handler = self.mapping.match(body)

            if handler is None:
                # XXX Is this the right behavior?
                abort(404)

            # Set up our response and request
            twiml_response = twiml.Response()
            twilio_request = TwilioMessageRequest(request)

            if match_obj is not None:
                kwargs = match_obj.groupdict()
            else:
                kwargs = {}

            try:
                handler(twilio_request, twiml_response, **kwargs)
                return unicode(twiml_response), 200, RESPONSE_HEADERS
            except:
                abort(500)

    def connect(self, pattern, flags=0):
        def _decorator(f):
            self.mapping.add_rule(pattern, f, flags=flags)
            @wraps(f)
            def wrapped(*args, **kwargs):
                return f(*args, **kwargs)
        return _decorator

    def run(self, debug=False):
        self.app.run(debug=debug)

    def __call__(self, *args, **kwargs):
        return self.app(*args, **kwargs)


strowger/switch.py
import re
from collections import namedtuple
from functools import wraps

from flask import Flask
from flask import abort
from flask import request
from twilio import twiml


Rule = namedtuple('Rule', 'pattern handler')
MediaItem = namedtuple('MediaItem', 'content_type url')

RESPONSE_HEADERS = {'Content-Type': 'application/xml'}


class Map(object):

    def __init__(self, default_handler=None):
        self.rules = []
        self.default_handler = default_handler

    def add_rule(self, pattern, handler, flags=0):
        compiled = re.compile(pattern, flags=flags)
        self.rules.append(Rule(compiled, handler))

    def match(self, body):
        for rule in self.rules:
            m = rule.pattern.match(body)
            if m is not None:
                return m, rule.handler

        return None, self.default_handler


class TwilioMessageRequest(object):

    def __init__(self, flask_request):
        self.from_number = flask_request.values['From']
        self.to_number = flask_request.values['To']
        self.message_sid = flask_request.values['MessageSid']
        self.account_sid = flask_request.values['AccountSid']
        self.message_body = flask_request.values['Body']
        self.media_count = int(flask_request.values['NumMedia'])
        self.media_items = []
        self.raw_values = flask_request.values
        self.flask_request = flask_request

        if self.media_count > 0:
            media_items = []
            for index in xrange(self.media_count):
                url = flask_request.values['MediaUrl{}'.format(index)]
                ctype = flask_request.values[
                    'MediaContentType{}'.format(index)
                ]
                media_items.append(MediaItem(ctype, url))

            self.media_items = media_items


class Switch(object):

    def __init__(self, name, request_path='/', methods=None,
                 default_handler=None, **kwargs):
        '''Create a Switch named `name`.

        Additional kwargs are passed to the underlying Flask application.
        '''
        self.app = Flask(name, **kwargs)
        self.mapping = Map(default_handler=default_handler)

        if methods is None:
            methods = ['GET', 'POST']

        @self.app.route(request_path, methods=methods)
        def twilio():
            body = request.values['Body']
            match_obj, handler = self.mapping.match(body)

            if handler is None:
                # XXX Is this the right behavior?
                abort(404)

            # Set up our response and request
            twiml_response = twiml.Response()
            twilio_request = TwilioMessageRequest(request)

            if match_obj is not None:
                kwargs = match_obj.groupdict()
            else:
                kwargs = {}

            response = handler(twilio_request, twiml_response, **kwargs)
            return unicode(response), 200, RESPONSE_HEADERS

    def connect(self, pattern, flags=0):
        def _decorator(f):
            self.mapping.add_rule(pattern, f, flags=flags)
            @wraps(f)
            def wrapped(*args, **kwargs):
                return f(*args, **kwargs)
        return _decorator

    def set_default_handler(self, handler):
        self.mapping.default_handler = handler

    def run(self, debug=False):
        self.app.run(debug=debug)

    def __call__(self, *args, **kwargs):
        return self.app(*args, **kwargs)
